### PR TITLE
Create `cardano-balance-tx` library shell. 🐚

### PR DESCRIPTION
**Pull request by jonathanknowles**
## Issue Number

ADP-2433

## Description

This PR:

- creates a new library `cardano-balance-tx` with a single top-level module (currently empty):
    ```patch
    + Cardano.Tx.Balance
    ```
- moves our existing balance-tx-specific `CoinSelection` module to the `cardano-balance-tx` library, and renames it, so that we have:
    ```patch
      Cardano.Tx.Balance
    + Cardano.Tx.Balance.Internal.CoinSelection
    ```
- renames the `cardano-wallet-coin-selection` library to `cardano-coin-selection`:
    ```patch
    - cardano-wallet-coin-selection
    + cardano-coin-selection
- simplifies the `cardano-coin-selection` library module hierarchy:
    ```patch
    - Cardano.Wallet.CoinSelection.Internal
    - Cardano.Wallet.CoinSelection.Internal.Balance
    - Cardano.Wallet.CoinSelection.Internal.Balance.Gen
    - Cardano.Wallet.CoinSelection.Internal.Collateral
    - Cardano.Wallet.CoinSelection.Internal.Context
    + Cardano.CoinSelection
    + Cardano.CoinSelection.Balance
    + Cardano.CoinSelection.Balance.Gen
    + Cardano.CoinSelection.Collateral
    + Cardano.CoinSelection.Context
    ```
- adjusts the dependencies of the `cardano-wallet` library, so that:
    - `cardano-wallet` **_does_** directly depend on `cardano-balance-tx`;
    - `cardano-wallet` **_does not_** directly depend on `cardano-coin-selection`.

### CODE REVIEW

**Comment by jonathanknowles** [comment]
bors try

File: cabal.project [hunk 1]
```diff
@@ -42,7 +42,8 @@ index-state: 2022-05-18T00:00:00Z
 with-compiler: ghc-8.10.7
 
 packages:
-    lib/coin-selection/
+    lib/balance-tx/
```
Comment (Anviking) [inline cabal.project:45]: 👍

File: lib/balance-tx/lib/Cardano/Tx/Balance.hs [hunk 2]
```diff
@@ -0,0 +1,7 @@
+-- |
+-- Copyright: © 2022 IOHK
+-- License: Apache-2.0
+--
+module Cardano.Tx.Balance
```
Comment (Anviking) [inline lib/balance-tx/lib/Cardano/Tx/Balance.hs:5]: I think I'd lean towards `Cardano.Write.Tx.Balance` and us renaming `Cardano.Wallet.Write.Tx` to `Cardano.Write.Tx`

**Review by Anviking — APPROVED** [review]

File: lib/balance-tx/lib/Cardano/Tx/Balance.hs (hunk 2 above)
Comment (Anviking) [reply lib/balance-tx/lib/Cardano/Tx/Balance.hs:5]: But can be reconciled/discussed later

Comment (jonathanknowles) [reply lib/balance-tx/lib/Cardano/Tx/Balance.hs:5]: > I think I'd lean towards Cardano.Write.Tx.Balance and us renaming Cardano.Wallet.Write.Tx to Cardano.Write.Tx
> But can be reconciled/discussed later

Definitely! We can revise this as we go later on. I'm happy to go with your suggestion, though I wonder if it still makes sense outside of the context of the wallet, where we do have a clear dichotomy between transactions that have been **_read_** and transactions that are to be **_written_**, and thus a desire to distinguish between these two use cases in the module hierarchy.

Conversely, in the `balance-tx` library, we (presumably) only ever deal with transactions that are intended to be "written" to the blockchain, so perhaps it's superfluous to specify it in the module hierarchy?

But on the other hand, if `balance-tx` is intended to exist within a **_family_** of libraries that collectively deal with reading and writing transactions, then nesting the main module within `Write` might make sense. 👍🏻

Comment (Anviking) [reply lib/balance-tx/lib/Cardano/Tx/Balance.hs:5]: Two arguments for including `Write`:
- It is a key concept that will be permeating the entire library.
- It is a pretty unique identifier (as opposed to `Cardano.Tx` ~ `Cardano.Api.Tx`)
    - Can use `as WriteTx` as qualified name as opposed to... overloaded `as Tx`?

We currently have
```haskell
import qualified Cardano.Wallet.Write.Tx as WriteTx
```
and I'm imagining:

```haskell
import qualified Cardano.Write.Tx as WriteTx -- either this
import qualified Cardano.Api.Write.Tx as WriteTx -- or this for cardano-api-ified version
-- ^ Motivation Cardano.Write.Tx.CardanoApi is ugly

import Cardano.Write.Tx.Balance (balanceTx)
```

**Pull request merged**